File: airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/MessageMetricsTracker.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.internal;

import io.airbyte.metrics.lib.MetricAttribute;
import io.airbyte.metrics.lib.MetricClient;
import io.airbyte.metrics.lib.MetricTags;
import io.airbyte.metrics.lib.OssMetricsRegistry;
import io.airbyte.protocol.models.AirbyteMessage.Type;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Helper to emit metrics around messages exchanged with connectors.
 */
public class MessageMetricsTracker {

  private static final MetricAttribute RECORD_ATTRIBUTE = new MetricAttribute(MetricTags.MESSAGE_TYPE, Type.RECORD.toString());
  private static final MetricAttribute STATE_ATTRIBUTE = new MetricAttribute(MetricTags.MESSAGE_TYPE, Type.STATE.toString());

  private final MetricClient metricClient;
  private final AtomicLong destRecordReadCount = new AtomicLong();
  private final AtomicLong destStateReadCount = new AtomicLong();
  private final AtomicLong destRecordSentCount = new AtomicLong();
  private final AtomicLong destStateSentCount = new AtomicLong();
  private final AtomicLong sourceRecordReadCount = new AtomicLong();
  private final AtomicLong sourceStateReadCount = new AtomicLong();
  private MetricAttribute connectionAttribute = null;

  public MessageMetricsTracker(final MetricClient metricClient) {
    this.metricClient = metricClient;
  }

  public void trackConnectionId(final UUID connectionId) {
    connectionAttribute = new MetricAttribute(MetricTags.CONNECTION_ID, connectionId.toString());
  }

  public void trackDestRead(final Type type) {
    if (type == Type.RECORD) {
      destRecordReadCount.incrementAndGet();
    } else if (type == Type.STATE) {
      destStateReadCount.incrementAndGet();
    }
  }

  public void trackDestSent(final Type type) {
    if (type == Type.RECORD) {
      destRecordSentCount.incrementAndGet();
    } else if (type == Type.STATE) {
      destStateSentCount.incrementAndGet();
    }
  }

  public void trackSourceRead(final Type type) {
    if (type == Type.RECORD) {
      sourceRecordReadCount.incrementAndGet();
    } else if (type == Type.STATE) {
      sourceStateReadCount.incrementAndGet();
    }
  }

  public void flushDestReadCountMetric() {
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_READ, destRecordReadCount, RECORD_ATTRIBUTE);
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_READ, destStateReadCount, STATE_ATTRIBUTE);
  }

  public void flushDestSentCountMetric() {
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_SENT, destRecordSentCount, RECORD_ATTRIBUTE);
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_SENT, destStateSentCount, STATE_ATTRIBUTE);
  }

  public void flushSourceReadCountMetric() {
    emitMetric(OssMetricsRegistry.WORKER_SOURCE_MESSAGE_READ, sourceRecordReadCount, RECORD_ATTRIBUTE);
    emitMetric(OssMetricsRegistry.WORKER_SOURCE_MESSAGE_READ, sourceStateReadCount, STATE_ATTRIBUTE);
  }

  private void emitMetric(final OssMetricsRegistry metric, final AtomicLong value, final MetricAttribute typeAttribute) {
    if (connectionAttribute != null) {
      metricClient.count(metric, value.getAndSet(0), connectionAttribute, typeAttribute);
    } else {
      metricClient.count(metric, value.getAndSet(0), typeAttribute);
    }
  }

}

File: airbyte-container-orchestrator/src/main/kotlin/io/airbyte/container/orchestrator/tracker/MessageMetricsTracker.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.tracker

import io.airbyte.metrics.MetricAttribute
import io.airbyte.metrics.MetricClient
import io.airbyte.metrics.OssMetricsRegistry
import io.airbyte.metrics.lib.MetricTags
import io.airbyte.protocol.models.v0.AirbyteMessage
import jakarta.inject.Singleton
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong

private val RECORD_ATTRIBUTE: MetricAttribute = MetricAttribute(MetricTags.MESSAGE_TYPE, AirbyteMessage.Type.RECORD.toString())
private val STATE_ATTRIBUTE: MetricAttribute = MetricAttribute(MetricTags.MESSAGE_TYPE, AirbyteMessage.Type.STATE.toString())

/**
 * Helper to emit metrics around messages exchanged with connectors.
 */
@Singleton
class MessageMetricsTracker(
  private val metricClient: MetricClient,
) {
  private val destRecordReadCount = AtomicLong()
  private val destStateReadCount = AtomicLong()
  private val destRecordSentCount = AtomicLong()
  private val destStateSentCount = AtomicLong()
  private val sourceRecordReadCount = AtomicLong()
  private val sourceStateReadCount = AtomicLong()
  private var connectionAttribute: MetricAttribute? = null

  fun trackConnectionId(connectionId: UUID?) {
    connectionId?.let {
      connectionAttribute = MetricAttribute(MetricTags.CONNECTION_ID, connectionId.toString())
    }
  }

  fun trackDestRead(type: AirbyteMessage.Type?) {
    if (type == AirbyteMessage.Type.RECORD) {
      destRecordReadCount.incrementAndGet()
    } else if (type == AirbyteMessage.Type.STATE) {
      destStateReadCount.incrementAndGet()
    }
  }

  fun trackDestSent(type: AirbyteMessage.Type?) {
    if (type == AirbyteMessage.Type.RECORD) {
      destRecordSentCount.incrementAndGet()
    } else if (type == AirbyteMessage.Type.STATE) {
      destStateSentCount.incrementAndGet()
    }
  }

  fun trackSourceRead(type: AirbyteMessage.Type?) {
    if (type == AirbyteMessage.Type.RECORD) {
      sourceRecordReadCount.incrementAndGet()
    } else if (type == AirbyteMessage.Type.STATE) {
      sourceStateReadCount.incrementAndGet()
    }
  }

  fun flushDestReadCountMetric() {
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_READ, destRecordReadCount, RECORD_ATTRIBUTE)
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_READ, destStateReadCount, STATE_ATTRIBUTE)
  }

  fun flushDestSentCountMetric() {
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_SENT, destRecordSentCount, RECORD_ATTRIBUTE)
    emitMetric(OssMetricsRegistry.WORKER_DESTINATION_MESSAGE_SENT, destStateSentCount, STATE_ATTRIBUTE)
  }

  fun flushSourceReadCountMetric() {
    emitMetric(OssMetricsRegistry.WORKER_SOURCE_MESSAGE_READ, sourceRecordReadCount, RECORD_ATTRIBUTE)
    emitMetric(OssMetricsRegistry.WORKER_SOURCE_MESSAGE_READ, sourceStateReadCount, STATE_ATTRIBUTE)
  }

  private fun emitMetric(
    metric: OssMetricsRegistry,
    value: AtomicLong,
    typeAttribute: MetricAttribute?,
  ) {
    if (connectionAttribute != null) {
      metricClient.count(metric, value.getAndSet(0), connectionAttribute, typeAttribute)
    } else {
      metricClient.count(metric, value.getAndSet(0), typeAttribute)
    }
  }
}
